kythe/go/serving/explore/explore.go

```diff
@@ -49,17 +57,108 @@ func (t *Tables) TypeHierarchy(ctx context.Context, req *epb.TypeHierarchyReques
 	return nil, nil
 }
 
-// Callers returns the (recursive) callers of a specified function, as a directed graph.
-// TODO: not yet implemented
+// Callers returns the callers of a specified function, as a directed graph.
 func (t *Tables) Callers(ctx context.Context, req *epb.CallersRequest) (*epb.CallersReply, error) {
-	return nil, nil
+	tickets := req.Tickets
+	if len(tickets) == 0 {
+		return nil, fmt.Errorf("missing input tickets: %v", &req)
+	}
+
+	// succMap maps nodes onto sets of successor nodes
+	succMap := make(map[string]map[string]bool)
+
+	// At the moment, this is our policy for missing data: if an input ticket has
+	// no record in the table, we don't include data for that ticket in the response.
+	// Other table access errors result in returning an error.
+	for _, ticket := range tickets {
+		succMap[ticket] = make(map[string]bool)
+		var callgraph srvpb.Callgraph
+		err := t.FunctionToCallers.Lookup(ctx, []byte(ticket), &callgraph)
+
+		if err != nil {
+			if err != table.ErrNoSuchKey {
+				return nil, fmt.Errorf("error looking up callers with ticket %q: %v", ticket, err)
+			}
+			continue // skip tickets with no mappings
+		}
+
+		if callgraph.Type != srvpb.Callgraph_CALLER {
+			return nil, fmt.Errorf("type of callgraph is not 'CALLER': %v", callgraph)
+		}
+
+		// TODO(jrtom): consider logging a warning if len(callgraph.Tickets) == 0
+		// (postprocessing should disallow this)
+		for _, predTicket := range callgraph.Tickets {
+			succMap[predTicket][ticket] = true
+		}
+	}
+
+	return &epb.CallersReply{Graph: convertSuccMapToGraph(succMap)}, nil
+}
+
+func convertSuccMapToGraph(succMap map[string]map[string]bool) *epb.Graph {
+	graph := &epb.Graph{
+		Nodes: make(map[string]*epb.GraphNode),
```

Comment:
Since we know the size: `make(map[string]*epb.GraphNode), len(succMap))`

Reply:
Well, we know it's at least that big, so I suppose it won't hurt to give the size hint, but there could be many more items in the values of "succMap" that don't have any successors and so never appear as keys.   Done.